bpf/kmesh/workload/include/authz.h

```diff
@@ -182,13 +185,312 @@ static int match_dst_ports(Istio__Security__Match *match, struct xdp_info *info,
+        }
+
+        IP6_COPY(target_addr.ip6, tuple_info->ipv6.daddr);
+        return matchIpv6(&rule_addr, &target_addr, preFixLen);
+    } else {
+        BPF_LOG(ERR, AUTH, "Unsupported address length: %u\n", addrInfo->len);
+    }
+
+    return UNMATCHED;
+}
+
+static inline int match_dst_ip(Istio__Security__Match *match, struct bpf_sock_tuple *tuple_info)
+{
+    void *dstPtrs = NULL;
+    void *notDstPtrs = NULL;
+    void *dstAddr = NULL;
+    void *notDstAddr = NULL;
+    Istio__Security__Address *dst = NULL;
+    Istio__Security__Address *notDst = NULL;
+    __u32 i = 0;
+
+    if (match->n_destination_ips == 0 && match->n_not_destination_ips == 0) {
+        BPF_LOG(DEBUG, AUTH, "no dstip configured, matching by default");
+        return MATCHED;
+    }
+
+    // match not_dstIPs
+    if (match->n_not_destination_ips != 0) {
+        notDstPtrs = KMESH_GET_PTR_VAL(match->not_destination_ips, void *);
+        if (!notDstPtrs) {
+            BPF_LOG(ERR, AUTH, "failed to retrieve not_dstips pointer\n");
+            return UNMATCHED;
+        }
+
+#pragma unroll
+        for (i = 0; i < MAX_MEMBER_NUM_PER_POLICY; i++) {
+            if (i >= match->n_not_destination_ips) {
+                break;
+            }
+
+            if (bpf_probe_read_kernel(&notDstAddr, sizeof(notDstAddr), &notDstPtrs[i]) != 0) {
+                BPF_LOG(ERR, AUTH, "failed to read notSrcAddr address at index %d", i);
+                continue;
+            }
+
+            notDst = (Istio__Security__Address *)KMESH_GET_PTR_VAL((void *)notDstAddr, Istio__Security__Address);
+            if (!notDst) {
+                continue;
+            }
+            if (matchIp(&notDst->address, notDst->length, tuple_info, TYPE_DSTIP) == MATCHED) {
+                return UNMATCHED;
+            }
+        }
+    }
+
+    if (match->n_destination_ips != 0) {
+        dstPtrs = KMESH_GET_PTR_VAL(match->destination_ips, void *);
+        if (!dstPtrs) {
+            BPF_LOG(ERR, AUTH, "failed to get dstips ptr\n");
+            return UNMATCHED;
+        }
+
+#pragma unroll
+        for (i = 0; i < MAX_MEMBER_NUM_PER_POLICY; i++) {
+            if (i >= match->n_destination_ips) {
+                break;
+            }
+
+            if (bpf_probe_read_kernel(&dstAddr, sizeof(dstAddr), &dstPtrs[i]) != 0) {
+                BPF_LOG(ERR, AUTH, "failed to read dst address at index %d", i);
+                continue;
+            }
+
+            dst = (Istio__Security__Address *)KMESH_GET_PTR_VAL((void *)dstAddr, Istio__Security__Address);
+            if (!dst) {
+                continue;
+            }
+            if (matchIp(&dst->address, dst->length, tuple_info, TYPE_DSTIP) == MATCHED) {
+                return MATCHED;
+            }
+        }
+    }
+    BPF_LOG(DEBUG, AUTH, "no matching dstip found, unmatched");
+    return UNMATCHED;
+}
+
+static inline int match_src_ip(Istio__Security__Match *match, struct bpf_sock_tuple *tuple_info)
```

Comment:
Can't the two functions match_dst_ip() and match_src_ip() be abstracted into one function? Most implementations are the same

Reply:
good catch @weli-l